modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -74,27 +86,45 @@ public TxManagerImpl(ReplicaService replicaService, LockManager lockManager, Hyb
         this.clock = clock;
     }
 
-    /** {@inheritDoc} */
     @Override
     public InternalTransaction begin() {
         return begin(false);
     }
 
-    /** {@inheritDoc} */
     @Override
     public InternalTransaction begin(boolean readOnly) {
         UUID txId = Timestamp.nextVersion().toUuid();
 
-        return readOnly ? new ReadOnlyTransactionImpl(this, txId, clock.now()) : new ReadWriteTransactionImpl(this, txId);
+        if (!readOnly) {
+            return new ReadWriteTransactionImpl(this, txId);
+        }
+
+        HybridTimestamp readTimestamp = clock.now();
+
+        readOnlyTxFutureByReadTs.compute(readTimestamp, (timestamp, readOnlyTxFuture) -> {
+            assert readOnlyTxFuture == null : "previous transaction has not completed yet: " + readTimestamp;
+
+            HybridTimestamp lowerBound = lowerBoundTsToStartNewReadOnlyTx.get();
+
+            if (lowerBound != null && readTimestamp.compareTo(lowerBound) <= 0) {
+                throw new IgniteInternalException(
+                        TX_READ_ONLY_CREATING_ERR,
+                        "Timestamp read-only transaction must be greater than the lower bound: [txTimestamp={}, lowerBound={}]",
+                        readTimestamp, lowerBound
+                );
```

Comment:
Consider the following scenario:

1. One thread (A) starts an RO TX, it is inside `readOnlyTxFutureByReadTs.compute()`, it already took the old lower bound value and validated the read TS (T1) successfully, but did not return from the closure yet
2. Another thread (B) updates the lower bound to a higher value, with which T1 is not valid anymore
3. Then A finishes the addition of the new RO TX, which is too old

Another scenario is that addition of a transaction with TS T1 begins in thread A, then thread B takes `getFutureAllReadOnlyTransactions()` for T2>T1 (and does not see that transaction for T1 yet), and then the addition of the transaction completes. The transaction might take longer than the transactions that were seen by `getFutureAllReadOnlyTransactions()`, so the result of `getFutureAllReadOnlyTransactions()` might complete before the transaction at T1 completes, which seems to break the contract.

Probably we need a synchronization coordinating updates of both the future map and the lower bound (and also the reads of the map that are used to build the compound future).

Reply:
You're right, for some reason I relied on the internal logic of the `ConcurrentSkipListMap` that it would be blocked on new keys, fixed it.